File: services/core/src/main/java/com/previopls/service/AuditService.java
package com.previopls.service;

import com.previopls.entity.AuditLog;
import com.previopls.entity.enums.AuditAction;
import com.previopls.repository.AuditLogRepository;
import com.previopls.security.RequestContext;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

/**
 * Registra eventos relevantes para auditoria (LGPD / segurança).
 *
 * Cada chamada cria uma nova transação ({@link Propagation#REQUIRES_NEW}) — o
 * audit é persistido mesmo que a transação de negócio principal rollback.
 *
 * Quando o chamador não informa o ator, o serviço preenche {@code actorId} e
 * {@code actorRole} a partir do JWT autenticado na requisição (inclusive o JWT
 * interno emitido pelo Gateway), para que a trilha responda "quem fez".
 */
@Service
public class AuditService {

    private final AuditLogRepository repository;

    public AuditService(AuditLogRepository repository) {
        this.repository = repository;
    }

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void log(AuditAction action,
                    UUID actorId,
                    String actorEmail,
                    String actorRole,
                    String entityType,
                    String entityId,
                    String details) {
        if (actorId == null || actorRole == null) {
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            if (auth != null && auth.isAuthenticated() && auth.getName() != null) {
                if (actorId == null) {
                    actorId = parseUuid(auth.getName());
                }
                if (actorRole == null) {
                    actorRole = auth.getAuthorities().stream()
                            .map(GrantedAuthority::getAuthority)
                            .filter(a -> a.startsWith("ROLE_"))
                            .map(a -> a.substring(5).toLowerCase())
                            .findFirst()
                            .orElse(null);
                }
            }
        }

        AuditLog log = new AuditLog();
        log.setAction(action);
        log.setActorId(actorId);
        log.setActorEmail(actorEmail);
        log.setActorRole(actorRole);
        log.setEntityType(entityType);
        log.setEntityId(entityId);
        log.setRequestId(RequestContext.currentRequestId());
        log.setRemoteIp(RequestContext.remoteIp());
        String ua = RequestContext.userAgent();
        if (ua != null && ua.length() > 255) ua = ua.substring(0, 255);
        log.setUserAgent(ua);
        log.setDetails(details);
        repository.save(log);
    }

    private static UUID parseUuid(String value) {
        try {
            return UUID.fromString(value);
        } catch (IllegalArgumentException e) {
            return null;
        }
    }

    public void logLoginSuccess(UUID userId, String email, String role) {
        log(AuditAction.LOGIN_SUCCESS, userId, email, role, "Usuario", userId.toString(), null);
    }

    public void logLoginFailed(String email, String reason) {
        log(AuditAction.LOGIN_FAILED, null, email, null, null, null, reason);
    }

    public void logLoginLocked(String email) {
        log(AuditAction.LOGIN_LOCKED, null, email, null, null, null, "lockout ativo");
    }
}

File: services/gateway/app/services/audit_service.py
from typing import Optional
from uuid import UUID

from fastapi import Request
from sqlalchemy.orm import Session

from app.core.logging import get_logger
from app.models import AuditAction, AuditLog


log = get_logger(__name__)


class AuditService:
    def __init__(self, db: Session) -> None:
        self.db = db

    def log_event(
        self,
        *,
        action: AuditAction,
        request: Optional[Request] = None,
        actor_id: Optional[UUID] = None,
        actor_email: Optional[str] = None,
        actor_role: Optional[str] = None,
        entity_type: Optional[str] = None,
        entity_id: Optional[str] = None,
        details: Optional[str] = None,
    ) -> AuditLog:
        request_id = None
        remote_ip = None
        user_agent = None
        if request is not None:
            request_id = getattr(request.state, "request_id", None)
            remote_ip = request.client.host if request.client else None
            user_agent = request.headers.get("User-Agent")
            if user_agent and len(user_agent) > 255:
                user_agent = user_agent[:255]
            principal = getattr(request.state, "principal", None)
            if principal and actor_id is None:
                actor_id = UUID(principal.user_id) if principal.user_id else None
                actor_role = principal.role.value
                actor_email = actor_email or principal.nome

        entry = AuditLog(
            action=action,
            actor_id=actor_id,
            actor_email=actor_email,
            actor_role=actor_role,
            entity_type=entity_type,
            entity_id=entity_id,
            request_id=request_id,
            remote_ip=remote_ip,
            user_agent=user_agent,
            details=details,
        )
        self.db.add(entry)
        self.db.flush()

        log.info(
            "audit",
            action=action.value,
            actor_id=str(actor_id) if actor_id else None,
            entity_id=entity_id,
        )
        return entry
